intermediate_source/named_tensor_tutorial.py

```diff
@@ -0,0 +1,518 @@
+# Accessors and Reduction
+# -----------------------
+#
+# One can use dimension names to refer to dimensions instead of the positional
+# dimension. These operations also propagate names. Indexing (basic and
+# advanced) has not been implemented yet but is on the roadmap. Using the `named_imgs`
+# tensor from above, we can do:
+
+output = named_imgs.sum(['C'])  # Perform a sum over the channel dimension
+print(output.names)
+
+img0 = named_imgs.select('N', 0)  # get one image
+print(img0.names)
+
+######################################################################
+# Name inference
+# --------------
+#
+# Names are propagated on operations in a two step process called **name inference**. It
+# works as follows:
+#
+# 1. **Check names**: an operator may check that certain dimensions must match.
+# 2. **Propagate names**: name inference computes and propagates output names to
+#    output tensors.
+#
+# Let's go through the very small example of adding 2 one-dim tensors with no
+# broadcasting.
+
+x = torch.randn(3, names=('X',))
+y = torch.randn(3)
+z = torch.randn(3, names=('Z',))
+
+# **Check names**: first, we will check whether the names of these two tensors
+# match. Two names match if and only if they are equal (string equality) or at
+# least one is ``None`` (``None``s are essentially a special wildcard name).
+# The only one of these three that will error, therefore, is ``x+z``:
+
+catch_error(lambda: x + z)
+
+# **Propagate names**: _unify_ the two names by returning the most refined name of
+# the two. With ``x + y``,  ``X`` is more specific than ``None``.
+
+print((x + y).names)
+
+######################################################################
+# Most name inference rules are straightforward but some of them (the dot
+# product ones) can have unexpected semantics. Let's go through a few more of
+# them.
+#
+# Broadcasting
+# ------------
+#
+# Named tensors do not change broadcasting behavior; they still broadcast by
+# position. However, when checking two dimensions for if they can be
+# broadcasted, the names of those dimensions must match. Two names match if and
+# only if they are equal (string equality), or if one is None.
+#
+# We do not support **automatic broadcasting** by names because the output
+# ordering is ambiguous and does not work well with unnamed dimensions. However,
+# we support **explicit broadcasting** by names, which is introduced in a later
+# section. The two examples below help clarify this.
+
+# Automatic broadcasting: expected to fail
+imgs = torch.randn(6, 6, 6, 6, names=('N', 'C', 'H', 'W'))
+per_batch_scale = torch.rand(6, names=('N',))
+catch_error(lambda: imgs * per_batch_scale)
+
+# Explicit broadcasting: the names check out and the more refined names are propagated.
+imgs = torch.randn(6, 6, 6, 6, names=('N', 'C', 'H', 'W'))
+per_batch_scale_4d = torch.rand(6, 1, 1, 1, names=('N', None, None, None))
+print((imgs * per_batch_scale_4d).names)

```

Comment:
None of these are "explicit broadcasting by names". They're both "automatic broadcasting" in the traditional (positional sense); the first example demonstrates where names prevent incorrect behavior and the second example demonstrates where names don't impede a correct result. Maybe we can clarify this point?